mui/dockwidgets/list_widget.py

```diff
@@ -0,0 +1,142 @@
+        allow_repeats: bool = True,
+        initial_row_count: int = 0,
+        validate_fun: Callable[[], None] = lambda: None,
+    ):
+        QWidget.__init__(self, parent)
+        self.validate_fun = validate_fun
+        self.possible_values = possible_values
+        self.allow_repeats = allow_repeats
+
+        self.row_layout = QVBoxLayout()
+
+        for i in range(initial_row_count):
+            self.add_row()
+
+        add_btn = QPushButton("+")
+        add_btn.clicked.connect(lambda: self.add_row())
+
+        layout = QVBoxLayout()
+        layout.addLayout(self.row_layout)
+        layout.addWidget(add_btn)
+
+        layout.setContentsMargins(0, 0, 0, 0)
+
+        self.setLayout(layout)
+
+    def add_row(self):
+        """Adds a new row to the current widget"""
+
+        if self.possible_values is None:
+            input = QLineEdit()
+            input.editingFinished.connect(lambda: self.validate_fun())
+        else:
+            input = QComboBox()
+
+            if self.allow_repeats:
+                input.addItems(self.possible_values)
+            else:
+                input.addItems(sorted(set(self.possible_values) - set(self.get_results())))
+
+            input.currentIndexChanged.connect(self._on_index_change)
+
+        btn = QPushButton("-")
+        btn.setMaximumWidth(20)
+
+        row = QHBoxLayout()
+        row.addWidget(input)
+        row.addWidget(btn)
+
+        btn.clicked.connect(lambda: self._on_remove_row([input, btn, row]))
+
+        self.row_layout.addLayout(row)
+
+        if not self.allow_repeats:
+            self._remove_repeats()
+
+    def _on_remove_row(self, elements_to_remove: List[QObject]):
+        """Remove a row from the list"""
+
+        for x in elements_to_remove:
+            x.setParent(None)
+            x.deleteLater()
+
+        if not self.allow_repeats:
+            self._remove_repeats()
+
+    def _on_index_change(self):
+        """Handle index change"""
+
+        self.validate_fun()
+
+        if not self.allow_repeats:
+            self._remove_repeats()
+
+    def _remove_repeats(self):
+        """Update the available options for each row to prevent repeats"""
+
+        assert self.possible_values is not None
+        assert not self.allow_repeats
+
+        not_selected = set(self.possible_values) - set(self.get_results())
+        for row in self.row_layout.children():
+            combobox: QComboBox = row.itemAt(0).widget()
+            curr_text = combobox.currentText()
+
+            # need to disable the event listener temporarily in order to
+            # prevent the same method being called recursively
+            combobox.currentIndexChanged.disconnect(self._on_index_change)
+
+            combobox.clear()
+            combobox.addItems(sorted(not_selected | {curr_text}))
+            combobox.setCurrentIndex(combobox.findText(curr_text))
+
+            combobox.currentIndexChanged.connect(self._on_index_change)
```

Comment:
This is probably fine for now since our list is small.

I think another way that _could_ work is to keep a list in this class that has all of the currently chosen options (`self.chosen_values`) and then when the user wants to change a single `QComboBox`, look at the `currentText` and make sure to still present it along with the `set(self.possible_values) - self.chosen_values`. I don't believe we need to change _all_ combo boxes? I might not be understanding how Qt is working at this point and/or if there's an event for "viewing" the available options...

The above probably describes what would happen if you turned your `self.get_results()` into the `self.chosen_values` variable. This would be a bit more accounting when the user is selecting, adding, and removing buttons though 😄 

Regardless, it's fine how it is and is working! Just some thoughts for the future maybe.

Edit: After some Googling, looks like you'd probably have to [override the class](https://stackoverflow.com/questions/21997846/populate-qcombobox-on-drop-down-or-user-click) to get what I'm describing. Let's leave that for a future performance improvement if it becomes an issue.

Reply:
Right, that sounds like a way more efficient solution! I can definitely look into it in the future.